### ISSUE TITLE
rubygems with native extensions reference /bin/sh directly

### ISSUE THREAD

**Issue by lilyball** [NixOS/nixpkgs#73310]
**Describe the bug**
When building a rubygem that has a native extension, it hardcodes `SHELL = /bin/sh`. If we can, we should fix this to point at the nixpkgs build shell instead.

It looks like the Makefile is generated by `extconf.rb`, and I'm pretty sure this is all just done as part of a `gem build` command, so I don't know how we can go in and patch it. Is there any way to cause `gem build` to pass additional make flags, or to synthesize a Makefile with a better `SHELL` to begin with?

**To Reproduce**
1. `nix-build --no-out-link --option sandbox true -A ruby.gems.ffi`. Access to `/bin/sh` is apparently allowed in the sandbox (at least on darwin, where I'm hitting this), but right now on macOS 10.15 it's failing because `/bin/sh` tries to execute `/bin/bash`.

**Expected behavior**
It should build the gem without touching `/bin/sh`.

**Output**
This comes from macOS 10.15 where `/bin/bash` is not allowed in the sandbox:

```
Building native extensions. This could take a while...
ERROR:  Error installing /nix/store/zq4aif72vx4y6sx56spyrjs0c538a608-ffi-1.10.0.gem:
        ERROR: Failed to build gem native extension.

    current directory: /nix/store/0gmx45yqhaj9bwpps3dxizblsxc4cq8x-ruby2.6.5-ffi-1.10.0/lib/ruby/gems/2.6.0/gems/ffi-1.10.0/ext/ffi_c
/nix/store/h1jmi5pnsjj8vlnmbzks547a3q83hdwb-ruby-2.6.5/bin/ruby -I /nix/store/h1jmi5pnsjj8vlnmbzks547a3q83hdwb-ruby-2.6.5/lib/ruby/2.6.0 -r ./siteconf20191112-74371-u5y091.rb extconf.rb
checking for ffi_call() in -lffi... yes
checking for ffi_closure_alloc()... yes
checking for ffi_raw_call()... yes
checking for ffi_prep_raw_closure()... yes
checking for shlwapi.h... no
checking for ruby/thread.h... yes
checking for rb_thread_blocking_region()... no
checking for rb_thread_call_with_gvl()... yes
checking for rb_thread_call_without_gvl()... yes
checking for ffi_prep_cif_var()... yes
creating extconf.h
creating Makefile

current directory: /nix/store/0gmx45yqhaj9bwpps3dxizblsxc4cq8x-ruby2.6.5-ffi-1.10.0/lib/ruby/gems/2.6.0/gems/ffi-1.10.0/ext/ffi_c
make "DESTDIR=" clean
Error opening /private/var/select/sh: Operation not permitted
Failed to exec /bin/bash as variant for /bin/sh (1: Operation not permitted).

current directory: /nix/store/0gmx45yqhaj9bwpps3dxizblsxc4cq8x-ruby2.6.5-ffi-1.10.0/lib/ruby/gems/2.6.0/gems/ffi-1.10.0/ext/ffi_c
make "DESTDIR="
Error opening /private/var/select/sh: Operation not permitted
Failed to exec /bin/bash as variant for /bin/sh (1: Operation not permitted).

make failed, exit code 1
```

**Metadata**
 - system: `"x86_64-darwin"`
 - host os: `Darwin 19.0.0, macOS 10.15.1`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.1`
 - channels(lilyball): `"nixpkgs-20.03pre199995.895874d2145"`
 - nixpkgs: `/Users/lilyball/.nix-defexpr/channels/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: ruby
# a list of nixos modules affected by the problem
module:
```

**Comment by lilyball** [comment]
The other possibility is that we're allowing access to `/bin/sh` for some reason, though I don't see any references to that in the sandbox. On macOS 10.15 `/bin/sh` actually access `/private/var/select/sh` which then causes it to execute `/bin/bash`.

**Comment by lilyball** [comment]
Come to think of it, this seems very plausible based on the error message above, as it's barfing on `/private/var/select/sh` rather than `/bin/sh` directly.

I still think it's a rubygems problem that it's running `/bin/sh` at all, but I'd love some pointers as to whether `/bin/sh` is supposed to be accessible to derivations.

**Comment by lilyball** [comment]
We do in fact allow `/bin/sh` in the profile, interesting. So this is specifically a macOS 10.15 issue.

**Comment by lilyball** [comment]
I still think this is problematic, but access to `/bin/sh` has been fixed for macOS 10.15 in general so it's no longer pressing. I'll close this out now and someone can reopen it if there's still a practical issue.
